include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+ * complex matrix-vector operations and free the user from having to
+ * create, set up and handle intermediate storage locations by hand.
+ *
+ * As an example consider the operation $(A+k\,B)\,C$, where $A$, $B$ and
+ * $C$ denote (possible different) matrices. In order to construct a
+ * LinearOperator <code>op</code> that stores the knowledge of this
+ * operation, one can write:
+ *
+ * @code
+ * dealii::SparseMatrix<double> A, B, C;
+ * const double k = ...;
+ *
+ * // Setup and assembly of matrices
+ *
+ * const auto op_a = linop(A);
+ * const auto op_b = linop(B);
+ * const auto op_c = linop(C);
+ *
+ * const auto op = (op_a + k * op_b) * op_c;
+ * @endcode
+ *
+ * @note: This class is only available if deal.II was configured with C++11
+ * support, i.e., if <code>DEAL_II_WITH_CXX11</code> is enabled during
+ * cmake configure.
+ *
+ * @author Luca Heltai, Matthias Maier, 2015
+ */
+template <typename Range, typename Domain> class LinearOperator
+{
+public:
+
+  /**
+   * Create an empty LinearOperator object. All <code>std::function</code>
+   * member objects are initialized with default variants that throw an
+   * exception upon invocation.
+   */
+  LinearOperator ()
+  {
+    vmult = [](Range &, const Domain &)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::vmult called"));
+    };
+
+    vmult_add = [](Range &, const Domain &)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::vmult_add called"));
+    };
+
+    Tvmult = [](Domain &, const Range &)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::Tvmult called"));
+    };
+
+    Tvmult_add = [](Domain &, const Range &)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::Tvmult_add called"));
+    };
+
+    reinit_range_vector = [](Range &, bool)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::reinit_range_vector method called"));
+    };
+
+    reinit_domain_vector = [](Domain &, bool)
+    {
+      Assert(false, ExcMessage("Uninitialized LinearOperator<Range, "
+                               "Domain>::reinit_domain_vector method called"));
+    };
+  }
+
+  /**
+   * Create a LinearOperator object from an object @p op for which the
+   * conversion function <code>linop</code> is defined.
+   */
+  template<typename Op>
+  LinearOperator (const Op &op)
+  {
+    *this = linop<Range, Domain, Op>(op);
+  }
+
+  /**
+   * Copy assignment operator for an object @p op for which the conversion
+   * function <code>linop</code> is defined.
+   */
+  template <typename Op>
+  LinearOperator<Range, Domain> &operator=(const Op &op)
+  {
+    return *this = linop<Range, Domain, Op>(op);
+  }
```

Comment:
Tested in ```tests/lac/linear_operator_02``` line 77ff.

The catch is that a templated operator=() does not prevent the default copy assignment (and copy constructor) variants to be automatically generated. And those are called by the ```return *this = (linear operator)``` statement.

I will add explicit statements
```
LinearOperator (const LinearOperator<Range, Domain> &) = default;
LinearOperator<Range, Domain> &operator=(const LinearOperator<Range, Domain> &) = default;
```
to make this a bit clearer.